horovod/tensorflow/mpi_ops.cc

```diff
@@ -480,5 +480,66 @@ Output
                `tensor` on root rank.
 )doc");
 
+class HorovodReducescatterOp : public AsyncOpKernel {
+public:
+  explicit HorovodReducescatterOp(OpKernelConstruction* context)
+      : AsyncOpKernel(context) {}
+
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) override {
+    OP_REQUIRES_OK_ASYNC(context, ConvertStatus(common::CheckInitialized()),
+                         done);
+
+    auto node_name = name();
+    auto device = GetDeviceID(context);
+    auto tensor = context->input(0);
+    // ReadyEvent makes sure input tensor is ready.  We cannot pre-allocate
+    // output for reducescatter, since shape of result is only known after all
+    // ranks make a request.
+    auto ready_event = std::shared_ptr<common::ReadyEvent>(RecordReadyEvent(context));
+    auto hvd_context = std::make_shared<TFOpContext>(context);
+    auto hvd_tensor = std::make_shared<TFTensor>(tensor);
+    auto enqueue_result = EnqueueTensorReducescatter(
+        hvd_context, hvd_tensor, ready_event, node_name, device,
+        [context, done](const common::Status& status) {
+          context->SetStatus(ConvertStatus(status));
+          done();
+        });
+    OP_REQUIRES_OK_ASYNC(context, ConvertStatus(enqueue_result), done);
+  }
+}; // namespace tensorflow
+
+REGISTER_KERNEL_BUILDER(Name("HorovodReducescatter").Device(DEVICE_CPU),
+                        HorovodReducescatterOp);
+#if HOROVOD_GPU_REDUCESCATTER
+REGISTER_KERNEL_BUILDER(Name("HorovodReducescatter").Device(DEVICE_GPU),
+                        HorovodReducescatterOp);
+#endif
+
+REGISTER_OP("HorovodReducescatter")
+    .Attr("T: {int32, int64, float16, float32, float64}")
+    .Input("tensor: T")
+    .Output("output: T")
+    .SetShapeFn([](shape_inference::InferenceContext* c) {
+      shape_inference::ShapeHandle output;
+      TF_RETURN_IF_ERROR(
+          c->ReplaceDim(c->input(0), 0, c->UnknownDim(), &output));
+      c->set_output(0, output);
+      return Status::OK();
+    })
+    .Doc(R"doc(
+Perform an MPI Reducescatter on a tensor. All other processes that do a
```

Comment:
I know this is consistent with the other docstrings here, but we should remove the reference to MPI at some point.

Reply:
Happy to update.